v8go.cc

```diff
@@ -148,79 +165,152 @@ IsolateHStatistics IsolationGetHeapStatistics(IsolatePtr ptr) {
                             hs.number_of_detached_contexts()};
 }
 
-/********** ObjectTemplate **********/
+/********** Template **********/
+
+#define LOCAL_TEMPLATE(ptr)                       \
+  m_template* ot = static_cast<m_template*>(ptr); \
+  Isolate* iso = ot->iso;                         \
+  Locker locker(iso);                             \
+  Isolate::Scope isolate_scope(iso);              \
+  HandleScope handle_scope(iso);                  \
+  Local<Template> tmpl = ot->ptr.Get(iso);
+
+void TemplateFree(TemplatePtr ptr) {
+  delete static_cast<m_template*>(ptr);
+}
+
+void TemplateSetValue(TemplatePtr ptr,
+                      const char* name,
+                      ValuePtr val_ptr,
+                      int attributes) {
+  LOCAL_TEMPLATE(ptr);
+
+  Local<String> prop_name =
+      String::NewFromUtf8(iso, name, NewStringType::kNormal).ToLocalChecked();
+  m_value* val = static_cast<m_value*>(val_ptr);
+  tmpl->Set(prop_name, val->ptr.Get(iso), (PropertyAttribute)attributes);
+}
+
+void TemplateSetTemplate(TemplatePtr ptr,
+                         const char* name,
+                         TemplatePtr obj_ptr,
+                         int attributes) {
+  LOCAL_TEMPLATE(ptr);
 
-#define LOCAL_OBJECT_TEMPLATE(ptr)                              \
-  m_object_template* ot = static_cast<m_object_template*>(ptr); \
-  Isolate* iso = ot->iso;                                       \
-  Locker locker(iso);                                           \
-  Isolate::Scope isolate_scope(iso);                            \
-  HandleScope handle_scope(iso);                                \
-  Local<ObjectTemplate> object_template = ot->ptr.Get(iso);
+  Local<String> prop_name =
+      String::NewFromUtf8(iso, name, NewStringType::kNormal).ToLocalChecked();
+  m_template* obj = static_cast<m_template*>(obj_ptr);
+  tmpl->Set(prop_name, obj->ptr.Get(iso), (PropertyAttribute)attributes);
+}
+
+/********** ObjectTemplate **********/
 
-ObjectTemplatePtr NewObjectTemplate(IsolatePtr iso_ptr) {
+TemplatePtr NewObjectTemplate(IsolatePtr iso_ptr) {
   Isolate* iso = static_cast<Isolate*>(iso_ptr);
   Locker locker(iso);
   Isolate::Scope isolate_scope(iso);
   HandleScope handle_scope(iso);
 
-  m_object_template* ot = new m_object_template;
+  m_template* ot = new m_template;
   ot->iso = iso;
   ot->ptr.Reset(iso, ObjectTemplate::New(iso));
-  return static_cast<ObjectTemplatePtr>(ot);
+  return static_cast<TemplatePtr>(ot);
 }
+/********** FunctionTemplate **********/
 
-void ObjectTemplateDispose(ObjectTemplatePtr ptr) {
-  delete static_cast<m_object_template*>(ptr);
+static void FunctionTemplateCallback(const FunctionCallbackInfo<Value>& info) {
```

Comment:
this is a little bit tricky and would love to see some comments in here as well!

Reply:
Added some comments, and actually simplified the data object. Hope it makes a little more sense now? (as much as it can with V8 C++ code)